third_party/eigen3/gpu_packet_math.patch

```diff
@@ -22,3 +22,120 @@
+ 
+-  template <int NPT, typename S, typename R, typename I_> KERNEL_FRIEND void internal::OuterReductionKernel(R, const S, I_, I_, typename S::CoeffReturnType*);
++  template <int NPT, typename S, typename R, typename I_, bool, bool> KERNEL_FRIEND void internal::OuterReductionKernel(R, const S, I_, I_, typename S::CoeffReturnType*);
+ #endif
+ 
+ #if defined(EIGEN_USE_SYCL)
+--- a/unsupported/Eigen/CXX11/src/Tensor/TensorReductionGpu.h
++++ b/unsupported/Eigen/CXX11/src/Tensor/TensorReductionGpu.h
+@@ -720,36 +720,61 @@
+   }
+ };
+ 
++
++#if GOOGLE_CUDA
++__shared__ char outer_reduction_shared_memory[];
++#endif
++
+ template <int NumPerThread, typename Self,
+-          typename Reducer, typename Index>
+-__global__ void OuterReductionKernel(Reducer reducer, const Self input, Index num_coeffs_to_reduce, Index num_preserved_coeffs,
+-                                     typename Self::CoeffReturnType* output) {
+-  const Index num_threads = blockDim.x * gridDim.x;
+-  const Index thread_id = blockIdx.x * blockDim.x + threadIdx.x;
++          typename Reducer, typename Index,
++          bool x64, bool use_shm>
++__global__ void OuterReductionKernel(Reducer reducer, const Self input, Index _num_coeffs_to_reduce, Index _num_preserved_coeffs,
++                                      typename Self::CoeffReturnType* output) {
++#if TENSORFLOW_USE_ROCM
++  HIP_DYNAMIC_SHARED(typename Self::CoeffReturnType, shared_memory);
++#else
++  typename Self::CoeffReturnType* shared_memory = reinterpret_cast<typename Self::CoeffReturnType*>(outer_reduction_shared_memory);
++#endif
++  typedef typename std::conditional<x64, Index, int>::type Ind;
++  Ind num_coeffs_to_reduce = _num_coeffs_to_reduce;
++  Ind num_preserved_coeffs = _num_preserved_coeffs;
++  const Ind num_threads = blockDim.x * gridDim.x;
++  const Ind thread_id = blockIdx.x * blockDim.x + threadIdx.x;
+   // Initialize the output values if they weren't initialized by the ReductionInitKernel
+   if (gridDim.x == 1) {
+-    for (Index i = thread_id; i < num_preserved_coeffs; i += num_threads) {
++    for (Ind i = thread_id; i < num_preserved_coeffs; i += num_threads) {
+       output[i] = reducer.initialize();
+     }
+     __syncthreads();
+   }
++  if(use_shm)
++    for(Ind i = threadIdx.x; i < num_preserved_coeffs; i += blockDim.x)
++     shared_memory[i] = reducer.initialize();
+ 
+   // Do the reduction.
+-  const Index max_iter = num_preserved_coeffs * divup<Index>(num_coeffs_to_reduce, NumPerThread);
+-  for (Index i = thread_id; i < max_iter; i += num_threads) {
+-    const Index input_col = i % num_preserved_coeffs;
+-    const Index input_row = (i / num_preserved_coeffs) * NumPerThread;
++  const Ind max_iter = num_preserved_coeffs * divup<Ind>(num_coeffs_to_reduce, NumPerThread);
++  for (Ind i = thread_id; i < max_iter; i += num_threads) {
++    Ind input_row = i / num_preserved_coeffs;
++    const Ind input_col = i - input_row*num_preserved_coeffs;
```

Comment:
Correct. The compiler may do it on its own, but, since a large part of the generated ISA is dedicated to performing i64/64 division (the hardware has no native support for integer division, so it gets emulated), it seems prudent to make sure.